File: upack/Pack.cs
using System;
using System.ComponentModel;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using Inedo.UPack;
using Inedo.UPack.Packaging;

namespace Inedo.ProGet.UPack
{
    [DisplayName("pack")]
    [Description("Creates a new ProGet universal package using specified metadata and source directory.")]
    public sealed class Pack : Command
    {
        [DisplayName("manifest")]
        [AlternateName("metadata")]
        [Description("Path of a valid upack.json metadata file.")]
        [ExtraArgument]
        public string Manifest { get; set; }

        [DisplayName("source")]
        [Description("Directory containing files to add to the package.")]
        [PositionalArgument(0)]
        [ExpandPath]
        public string SourceDirectory { get; set; }

        [DisplayName("targetDirectory")]
        [Description("Directory where the .upack file will be created. If not specified, the current working directory is used.")]
        [ExtraArgument]
        [ExpandPath]
        public string TargetDirectory { get; set; }

        [DisplayName("group")]
        [Description("Package group. If metadata file is provided, value will be ignored.")]
        [ExtraArgument]
        public string Group { get; set; }

        [DisplayName("name")]
        [Description("Package name. If metadata file is provided, value will be ignored.")]
        [ExtraArgument]
        public string Name { get; set; }

        [DisplayName("version")]
        [Description("Package version. If metadata file is provided, value will be ignored.")]
        [ExtraArgument]
        public string Version { get; set; }

        [DisplayName("title")]
        [Description("Package title. If metadata file is provided, value will be ignored.")]
        [ExtraArgument]
        public string Title { get; set; }

        [DisplayName("description")]
        [Description("Package description. If metadata file is provided, value will be ignored.")]
        [ExtraArgument]
        public string PackageDescription { get; set; }

        [DisplayName("icon")]
        [Description("Icon absolute Url. If metadata file is provided, value will be ignored.")]
        [ExtraArgument]
        public string IconUrl { get; set; }

        public override async Task<int> RunAsync(CancellationToken cancellationToken)
        {
            UniversalPackageMetadata info;

            if (string.IsNullOrWhiteSpace(this.Manifest))
            {
                info = new UniversalPackageMetadata
                {
                    Group = this.Group,
                    Name = this.Name,
                    Version = UniversalPackageVersion.TryParse(this.Version),
                    Title = this.Title,
                    Description = this.PackageDescription,
                    Icon = this.IconUrl
                };
            }
            else
            {
                if (!File.Exists(this.Manifest))
                {
                    Console.Error.WriteLine($"The manifest file '{this.Manifest}' does not exist.");
                    return 2;
                }

                using (var metadataStream = File.OpenRead(this.Manifest))
                {
                    info = await ReadManifestAsync(metadataStream);
                }
            }

            var error = ValidateManifest(info);
            if (error != null)
            {
                Console.Error.WriteLine("Invalid {0}: {1}", string.IsNullOrWhiteSpace(this.Manifest) ? "parameters" : "upack.json", error);
                return 2;
            }

            PrintManifest(info);

            if (!Directory.Exists(this.SourceDirectory))
            {
                Console.Error.WriteLine($"The source directory '{this.SourceDirectory}' does not exist.");
                return 2;
            }

            string relativePackageFileName = $"{info.Name}-{info.Version.Major}.{info.Version.Minor}.{info.Version.Patch}.upack";
            string targetFileName = Path.Combine(this.TargetDirectory ?? Environment.CurrentDirectory, relativePackageFileName);

            if (File.Exists(Path.Combine(this.SourceDirectory, relativePackageFileName)))
            {
                Console.Error.WriteLine("Warning: output file already exists in source directory and may be included inadvertently in the package contents.");
            }

            string tmpPath = Path.GetTempFileName();
            using (var builder = new UniversalPackageBuilder(tmpPath, info))
            {
                await builder.AddContentsAsync(
                    this.SourceDirectory, 
                    "/", 
                    true, 
                    s => string.IsNullOrWhiteSpace(this.Manifest) || !string.Equals(s, "upack.json", StringComparison.OrdinalIgnoreCase), 
                    cancellationToken
                );
            }

            Directory.CreateDirectory(Path.GetDirectoryName(targetFileName));
            File.Delete(targetFileName);
            File.Move(tmpPath, targetFileName);

            return 0;
        }
    }
}

File: pack.go
package main

import (
	"archive/zip"
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
)

type Pack struct {
	Manifest        string
	Metadata        UniversalPackageMetadata
	SourceDirectory string
	TargetDirectory string
}

func (*Pack) Name() string { return "pack" }
func (*Pack) Description() string {
	return "Creates a new ProGet universal package using specified metadata and source directory."
}

func (p *Pack) Help() string  { return defaultCommandHelp(p) }
func (p *Pack) Usage() string { return defaultCommandUsage(p) }

func (*Pack) PositionalArguments() []PositionalArgument {
	return []PositionalArgument{
		{
			Name:        "source",
			Description: "Directory containing files to add to the package.",
			Index:       0,
			TrySetValue: trySetPathValue("source", func(cmd Command) *string {
				return &cmd.(*Pack).SourceDirectory
			}),
		},
	}
}
func (*Pack) ExtraArguments() []ExtraArgument {
	return []ExtraArgument{
		{
			Name:        "manifest",
			Alias:       []string{"metadata"},
			Description: "Path of a valid upack.json metadata file.",
			TrySetValue: trySetPathValue("manifest", func(cmd Command) *string {
				return &cmd.(*Pack).Manifest
			}),
		},
		{
			Name:        "targetDirectory",
			Description: "Directory where the .upack file will be created. If not specified, the current working directory is used.",
			TrySetValue: trySetPathValue("targetDirectory", func(cmd Command) *string {
				return &cmd.(*Pack).TargetDirectory
			}),
		},
		{
			Name:        "group",
			Description: "Package group. If metadata file is provided, value will be ignored.",
			TrySetValue: trySetStringValue("group", func(cmd Command) *string {
				return &cmd.(*Pack).Metadata.Group
			}),
		},
		{
			Name:        "name",
			Description: "Package name. If metadata file is provided, value will be ignored.",
			TrySetValue: trySetStringValue("name", func(cmd Command) *string {
				return &cmd.(*Pack).Metadata.Name
			}),
		},
		{
			Name:        "version",
			Description: "Package version. If metadata file is provided, value will be ignored.",
			TrySetValue: trySetStringValue("version", func(cmd Command) *string {
				return &cmd.(*Pack).Metadata.Version
			}),
		},
		{
			Name:        "title",
			Description: "Package title. If metadata file is provided, value will be ignored.",
			TrySetValue: trySetStringValue("title", func(cmd Command) *string {
				return &cmd.(*Pack).Metadata.Title
			}),
		},
		{
			Name:        "description",
			Description: "Package description. If metadata file is provided, value will be ignored.",
			TrySetValue: trySetStringValue("description", func(cmd Command) *string {
				return &cmd.(*Pack).Metadata.Description
			}),
		},
		{
			Name:        "icon",
			Description: "Icon absolute URL. If metadata file is provided, value will be ignored.",
			TrySetValue: trySetStringValue("icon", func(cmd Command) *string {
				return &cmd.(*Pack).Metadata.IconURL
			}),
		},
	}
}

func (p *Pack) Run() int {
	if p.TargetDirectory == "" {
		p.TargetDirectory = "."
	}

	info := &p.Metadata
	if p.Manifest != "" {
		var err error
		info, err = p.ReadManifest()
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			return 1
		}
	}

	err := ValidateManifest(info)
	if err != nil {
		thing := "upack.json:"
		if strings.TrimSpace(p.Manifest) == "" {
			thing = "parameters:"
		}
		fmt.Fprintln(os.Stderr, "Invalid", thing, err)
		return 2
	}

	PrintManifest(info)

	fi, err := os.Stat(p.SourceDirectory)
	if os.IsNotExist(err) || (err == nil && !fi.IsDir()) {
		fmt.Fprintf(os.Stderr, "The source directory '%s' does not exist.\n", p.SourceDirectory)
		return 2
	} else if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}

	_, err = os.Stat(filepath.Join(p.SourceDirectory, info.Name+"-"+info.BareVersion()+".upack"))
	if err == nil {
		fmt.Fprintln(os.Stderr, "Warning: output file already exists in source directory and may be included inadvertently in the package contents.")
	} else if !os.IsNotExist(err) {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}

	targetFileName := filepath.Join(p.TargetDirectory, info.Name+"-"+info.BareVersion()+".upack")
	tmpFile, err := ioutil.TempFile("", "upack")
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	tmpPath := tmpFile.Name()
	defer func() {
		if tmpFile != nil {
			_ = tmpFile.Close()
			_ = os.Remove(tmpPath)
		}
	}()

	zipFile := zip.NewWriter(tmpFile)

	if p.Manifest != "" {
		err = CreateEntryFromFile(zipFile, p.Manifest, "upack.json")
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			return 1
		}
	} else {
		var buf bytes.Buffer
		err = json.NewEncoder(&buf).Encode(&p.Metadata)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			return 1
		}

		err = CreateEntryFromStream(zipFile, &buf, "upack.json")
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			return 1
		}
	}

	err = AddDirectory(zipFile, p.SourceDirectory, "package/")
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}

	err = zipFile.Close()
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}

	err = os.MkdirAll(filepath.Dir(targetFileName), 0755)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	err = os.Remove(targetFileName)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	err = tmpFile.Close()
	tmpFile = nil
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	err = os.Rename(targetFileName, tmpPath)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}

	return 0
}

func (p *Pack) ReadManifest() (*UniversalPackageMetadata, error) {
	f, err := os.Open(p.Manifest)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	return ReadManifest(f)
}
